File: DataStructures/AssociativeArrays/DynamicArray.js
const fs = require('fs')

class DynamicArray {
    /**
     * Creates an empty array of size 8.
     */
    constructor() {
        this.arr = Array(8)
        this.size = 0
    }

    /**
     * Checks if array is empty
     */
    isEmpty() {
        return this.size === 0
    }

    /**
     * Adds item to end of array.
     */
    push(item) {
        if (this.size == this.arr.length) this.resize(this.arr.length * 2)
        this.arr[this.size] = item
        this.size++
    }

    /**
     * Removes item from end of array
     */
    pop() {
        if (this.size === 0) throw new Error("Array is empty.")
        const item = this.arr[this.size - 1]
        this.size--
        this.arr[this.size] = null
        if (this.size > 0 && this.size === this.arr.length / 4) this.resize(this.arr.length / 2)
        return item
    }

    /**
     * Resizes array
     */
    resize(newSize) {
        const newArray = Array(newSize)
        for (let i = 0; i < this.size; i++) {
            newArray[i] = this.arr[i]
        }
        this.arr = newArray
    }

    /**
     * Returns item at given index
     */
    itemAt(index) {
        if (this.isEmpty() || index < 0 || index > this.size - 1) throw new Error("Index out of range")
        return this.arr[index]
    }
}

/* Test Section */
function main() {
    fs.readFile(__dirname + '/Tests/arrayTest.txt', 'utf8',
        function(err, data) {
            if (err) throw err
            const array = new DynamicArray()
            const testData = data.split('\n')
            for (let i = 0; i < testData.length; i++) {
                let item = testData[i]
                if (item === '-') {
                    array.pop()
                } else {
                    array.push(item)
                }
            }
            console.log(`Size of array: ${array.size}`) // should be -> 2
            console.log(`Array length: ${array.arr.length}`) // should be -> 4
        }
    )
}

main()

File: DataStructures/AssociativeArrays/DynamicArray.java
package DataStructures.AssociativeArrays;

import java.util.NoSuchElementException;

/**
 * Whe creating an array, Specify type of array using <> notation.
 * Ex) 
 * DynamicArray<String> = new DynamicArray<>(); for a String array.
 * DynamicArray<Integer> = new DynamicArray<>(); for an Integer array.
 */
public class DynamicArray<Item> {

    private int INITIAL_SIZE = 8;
    private Item[] arr;
    private int size;

    /**
     * Creates an empty array with size 8.
     */
    public DynamicArray() {
        arr = (Item[]) new Object[INITIAL_SIZE];
        size = 0;
    }

    /**
     * Checks if array is empty.
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * Gets item at specified index.
     */
    public Item itemAt(int index) {
        if (index < 0 || isEmpty() || index > size - 1) throw new IndexOutOfBoundsException();
        return arr[index];
    }

    /**
     * Adds an item to the array.
     */
    public void push(Item item) {
        if (size == arr.length) resize(arr.length * 2);
        arr[size++] = item;
    }

    /**
     * Removes item from end of array.
     */
    public Item pop() {
        if (isEmpty()) throw new NoSuchElementException("Array is empty.");
        Item item = arr[size - 1];
        arr[--size] = null; // Avoid loitering
        if (size > 0 && size == arr.length / 4) resize(arr.length / 2);
        return item;
    }

    /**
     * Resizes array. Copies all items from old array to new one.
     */
    private void resize(int newSize) {
        Item[] newArray = (Item[]) new Object[newSize];
        for (int i = 0; i < size; i++) {
            newArray[i] = arr[i];
        }
        arr = newArray;
    }

    /* Test Client */
    public static void main(String[] args) {
        
    }
}
